public/app/features/alerting/unified/components/settings/SettingsContext.tsx

```diff
@@ -0,0 +1,181 @@
+  enableAlertmanager: (uid: string) => void;
+  disableAlertmanager: (uid: string) => void;
+
+  // for updating or resetting the configuration for an Alertmanager
+  updateAlertmanagerSettings: (name: string, oldConfig: string, newConfig: string) => void;
+  resetAlertmanagerSettings: (name: string) => void;
+}
+
+const SettingsContext = React.createContext<Context | undefined>(undefined);
+const isInternalAlertmanager = (uid: string) => uid === GRAFANA_RULES_SOURCE_NAME;
+
+export const SettingsProvider = (props: PropsWithChildren) => {
+  // this list will keep track of Alertmanager UIDs (including internal) that are interested in receiving alert instances
+  // this will be used to infer the correct "delivery mode" and update the correct list of datasources with "wantsAlertsReceived"
+  let interestedAlertmanagers: string[] = [];
+
+  const { currentData: deliverySettings, isLoading: isLoadingDeliverySettings } =
+    alertmanagerApi.endpoints.getExternalAlertmanagerConfig.useQuery();
+
+  const [updateDeliverySettings, updateDeliverySettingsState] =
+    alertmanagerApi.endpoints.saveExternalAlertmanagersConfig.useMutation();
+  const [enableGrafanaManagedAlerts, enableOrDisableHandlingGrafanaManagedAlertsState] =
+    enableOrDisableHandlingGrafanaManagedAlerts();
+
+  const externalAlertmanagersWithStatus = useExternalDataSourceAlertmanagers();
+
+  const interestedInternal = USING_INTERNAL_ALERTMANAGER_SETTINGS.some(
+    (choice) => deliverySettings?.alertmanagersChoice === choice
+  );
+  if (interestedInternal) {
+    interestedAlertmanagers.push(GRAFANA_RULES_SOURCE_NAME);
+  }
+
+  externalAlertmanagersWithStatus
+    .filter((dataSource) => isAlertmanagerDataSourceInterestedInAlerts(dataSource.dataSourceSettings))
+    .forEach((alertmanager) => {
+      interestedAlertmanagers.push(alertmanager.dataSourceSettings.uid);
+    });
+
+  const enableAlertmanager = (uid: string) => {
+    const updatedInterestedAlertmanagers = union([uid], interestedAlertmanagers); // union will give us a unique array of uids
+    const newDeliveryMode = determineDeliveryMode(updatedInterestedAlertmanagers);
+    if (newDeliveryMode === null) {
+      return;
+    }
+
+    if (newDeliveryMode !== deliverySettings?.alertmanagersChoice) {
+      updateDeliverySettings({ alertmanagersChoice: newDeliveryMode });
+    }
+
+    if (!isInternalAlertmanager(uid)) {
+      enableGrafanaManagedAlerts(uid, true);
+    }
+  };
+
+  const disableAlertmanager = (uid: string) => {
+    const updatedInterestedAlertmanagers = without(interestedAlertmanagers, uid);
+    const newDeliveryMode = determineDeliveryMode(updatedInterestedAlertmanagers);
+    if (newDeliveryMode === null) {
+      return;
+    }
+
+    if (newDeliveryMode !== deliverySettings?.alertmanagersChoice) {
+      updateDeliverySettings({ alertmanagersChoice: newDeliveryMode });
+    }
+
+    if (!isInternalAlertmanager(uid)) {
+      enableGrafanaManagedAlerts(uid, false);
```

Comment:
I think the `enable` word and passing `true/false` argument can be misleading
Can we change the name to `set` or `toggle` or have two separate `enable` and `disable` functions?

Reply:
Good shout!